# NYC Airbnb Data Cleaning & Analysis

**Tools:** Python, Pandas, NumPy

**Dataset:** New York City Airbnb Open Data

**Project Type:** Data Cleaning & Exploratory Data Analysis

## Project Objective

The objective of this project is to clean and analyze the NYC Airbnb dataset using Python, Pandas, and NumPy.

The project focuses on identifying missing values, handling invalid data, detecting outliers, validating data quality, and extracting useful insights from Airbnb listings.

## 1. Data Inspection

Before cleaning the dataset, the data was inspected to understand its structure, size, columns, data types, and overall quality.

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("AB_NYC_2019.csv")

In [3]:
df.head()

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
0,2539,Clean & quiet apt home by the park,2787,John,Brooklyn,Kensington,40.64749,-73.97237,Private room,149,1,9,2018-10-19,0.21,6,365
1,2595,Skylit Midtown Castle,2845,Jennifer,Manhattan,Midtown,40.75362,-73.98377,Entire home/apt,225,1,45,2019-05-21,0.38,2,355
2,3647,THE VILLAGE OF HARLEM....NEW YORK !,4632,Elisabeth,Manhattan,Harlem,40.80902,-73.94190,Private room,150,3,0,NaN,NaN,1,365
3,3831,Cozy Entire Floor of Brownstone,4869,LisaRoxanne,Brooklyn,Clinton Hill,40.68514,-73.95976,Entire home/apt,89,1,270,2019-07-05,4.64,1,194
4,5022,Entire Apt: Spacious Studio/Loft by central park,7192,Laura,Manhattan,East Harlem,40.79851,-73.94399,Entire home/apt,80,10,9,2018-11-19,0.10,1,0


In [4]:
df.shape

(48895, 16)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

## 2. Missing Values & Duplicate Check

Missing values and duplicate records were checked before performing data cleaning.

In [6]:
df.isnull().sum()

,0
id,0
name,16
host_id,0
host_name,21
neighbourhood_group,0
neighbourhood,0
latitude,0
longitude,0
room_type,0
price,0


In [7]:
df.duplicated().sum()

np.int64(0)

In [8]:
df.describe()

,id,host_id,latitude,longitude,price,minimum_nights,number_of_reviews,reviews_per_month,calculated_host_listings_count,availability_365
count,4.889500e+04,4.889500e+04,48895.000000,48895.000000,48895.000000,48895.000000,48895.000000,38843.000000,48895.000000,48895.000000
mean,1.901714e+07,6.762001e+07,40.728949,-73.952170,152.720687,7.029962,23.274466,1.373221,7.143982,112.781327
std,1.098311e+07,7.861097e+07,0.054530,0.046157,240.154170,20.510550,44.550582,1.680442,32.952519,131.622289
min,2.539000e+03,2.438000e+03,40.499790,-74.244420,0.000000,1.000000,0.000000,0.010000,1.000000,0.000000
25%,9.471945e+06,7.822033e+06,40.690100,-73.983070,69.000000,1.000000,1.000000,0.190000,1.000000,0.000000
50%,1.967728e+07,3.079382e+07,40.723070,-73.955680,106.000000,3.000000,5.000000,0.720000,1.000000,45.000000
75%,2.915218e+07,1.074344e+08,40.763115,-73.936275,175.000000,5.000000,24.000000,2.020000,2.000000,227.000000
max,3.648724e+07,2.743213e+08,40.913060,-73.712990,10000.000000,1250.000000,629.000000,58.500000,327.000000,365.000000


## 3. Data Cleaning

The dataset was cleaned by handling missing values, removing invalid records, correcting data types, and creating additional fields for analysis.

### 3.1 Handling Invalid Prices

Listings with a price of $0 were identified as invalid for price analysis and removed from the dataset.

In [9]:
(df["price"] == 0).sum()

np.int64(11)

In [10]:
df = df[df["price"] > 0]

In [11]:
(df["price"] == 0).sum()

np.int64(0)

In [12]:
df.dtypes

,0
id,int64
name,object
host_id,int64
host_name,object
neighbourhood_group,object
neighbourhood,object
latitude,float64
longitude,float64
room_type,object
price,int64


### 3.2 Converting Date Data Type

The `last_review` column was converted from an object data type to a datetime format for accurate date handling.

In [13]:
df["last_review"] = pd.to_datetime(df["last_review"])

In [14]:
df["last_review"].dtype

dtype('<M8[ns]')

### 3.3 Handling Missing Text Values

Missing values in `name` and `host_name` were replaced with "Unknown" instead of removing the affected listings.

In [55]:
df["name"] = df["name"].fillna("Unknown")
df["host_name"] = df["host_name"].fillna("Unknown")

In [16]:
df[["name", "host_name"]].isnull().sum()

,0
name,0
host_name,0


### 3.4 Handling Missing Review Data

Missing values in `reviews_per_month` were replaced with 0 because these listings had no recorded monthly review activity.

In [17]:
df["reviews_per_month"] = df["reviews_per_month"].fillna(0)

In [18]:
df["reviews_per_month"].isnull().sum()

np.int64(0)

## 4. Outlier Detection

The Interquartile Range (IQR) method was used to identify unusually high Airbnb prices. Outliers were flagged instead of being removed because high-priced listings can represent genuine properties.

In [19]:
df["has_review"] = np.where(df["last_review"].isna(), "No Review", "Reviewed")

In [20]:
df["has_review"].value_counts()

,count
has_review,
Reviewed,38833
No Review,10051


In [21]:
(df[["price", "minimum_nights", "number_of_reviews",
     "reviews_per_month", "calculated_host_listings_count",
     "availability_365"]] < 0).sum()

,0
price,0
minimum_nights,0
number_of_reviews,0
reviews_per_month,0
calculated_host_listings_count,0
availability_365,0


In [22]:
df.columns

Index(['id', 'name', 'host_id', 'host_name', 'neighbourhood_group',
       'neighbourhood', 'latitude', 'longitude', 'room_type', 'price',
       'minimum_nights', 'number_of_reviews', 'last_review',
       'reviews_per_month', 'calculated_host_listings_count',
       'availability_365', 'has_review'],
      dtype='object')

In [23]:
df["room_type"].value_counts()

,count
room_type,
Entire home/apt,25407
Private room,22319
Shared room,1158


In [24]:
df["neighbourhood_group"].value_counts()

,count
neighbourhood_group,
Manhattan,21660
Brooklyn,20095
Queens,5666
Bronx,1090
Staten Island,373


In [25]:
df["neighbourhood"].nunique()

221

In [26]:
df["neighbourhood"].isnull().sum()

np.int64(0)

In [27]:
(df["price"] > 1000).sum()

np.int64(239)

In [28]:
df[df["price"] > 1000][["name", "neighbourhood_group", "room_type", "price"]].head(10)

,name,neighbourhood_group,room_type,price
496,Luxury 2Bed/2.5Bath Central Park View,Manhattan,Entire home/apt,2000
762,6 Bedroom Landmark West Village Townhouse,Manhattan,Entire home/apt,1300
946,Beautiful 3 bedroom in Manhattan,Manhattan,Private room,3000
1105,$ (Phone number hidden by Airbnb) weeks - room f,Manhattan,Private room,1300
1480,Lux 2Bed/2.5Bath Central Park Views,Manhattan,Entire home/apt,2000
1862,"Sunny, Family-Friendly 2 Bedroom",Brooklyn,Entire home/apt,4000
2018,Architecturally Stunning Former Synagogue!,Manhattan,Entire home/apt,2500
2236,Beautiful eco triplex w/green roof. Free yoga/...,Brooklyn,Entire home/apt,1395
2520,"Marcel the Shell with Shoes On or Off, Whatever",Brooklyn,Private room,1500
2523,West Village Penthouse-terrace/view,Manhattan,Entire home/apt,1899


### 4.1 IQR Method for Price

The first quartile (Q1), third quartile (Q3), and IQR were calculated to determine the upper price boundary.

In [29]:
Q1 = df["price"].quantile(0.25)
Q3 = df["price"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)
print("Lower Bound:", lower_bound)
print("Upper Bound:", upper_bound)

Q1: 69.0
Q3: 175.0
IQR: 106.0
Lower Bound: -90.0
Upper Bound: 334.0


In [30]:
(df["price"] > upper_bound).sum()

np.int64(2972)

### 4.2 Flagging Price Outliers

Listings above the calculated upper price boundary were flagged as outliers while keeping the original price data unchanged.

In [31]:
df["price_status"] = np.where(
    df["price"] > upper_bound,
    "Outlier",
    "Normal"
)

In [32]:
df["price_status"].value_counts()

,count
price_status,
Normal,45912
Outlier,2972


In [33]:
df.isnull().sum()

,0
id,0
name,0
host_id,0
host_name,0
neighbourhood_group,0
neighbourhood,0
latitude,0
longitude,0
room_type,0
price,0


In [34]:
df.shape

(48884, 18)

In [35]:
df.to_csv("cleaned_airbnb.csv", index=False)

In [36]:
import os

os.path.exists("cleaned_airbnb.csv")

True

In [37]:
df.head()

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365,has_review,price_status
0,2539,Clean & quiet apt home by the park,2787,John,Brooklyn,Kensington,40.64749,-73.97237,Private room,149,1,9,2018-10-19,0.21,6,365,Reviewed,Normal
1,2595,Skylit Midtown Castle,2845,Jennifer,Manhattan,Midtown,40.75362,-73.98377,Entire home/apt,225,1,45,2019-05-21,0.38,2,355,Reviewed,Normal
2,3647,THE VILLAGE OF HARLEM....NEW YORK !,4632,Elisabeth,Manhattan,Harlem,40.80902,-73.94190,Private room,150,3,0,NaT,0.00,1,365,No Review,Normal
3,3831,Cozy Entire Floor of Brownstone,4869,LisaRoxanne,Brooklyn,Clinton Hill,40.68514,-73.95976,Entire home/apt,89,1,270,2019-07-05,4.64,1,194,Reviewed,Normal
4,5022,Entire Apt: Spacious Studio/Loft by central park,7192,Laura,Manhattan,East Harlem,40.79851,-73.94399,Entire home/apt,80,10,9,2018-11-19,0.10,1,0,Reviewed,Normal


In [38]:
df.shape

(48884, 18)

In [39]:
df[["price", "minimum_nights", "number_of_reviews",
    "reviews_per_month", "availability_365"]].describe()

,price,minimum_nights,number_of_reviews,reviews_per_month,availability_365
count,48884.000000,48884.000000,48884.000000,48884.000000,48884.000000
mean,152.755053,7.029887,23.271991,1.090800,112.779498
std,240.170260,20.512224,44.551331,1.597213,131.627271
min,10.000000,1.000000,0.000000,0.000000,0.000000
25%,69.000000,1.000000,1.000000,0.040000,0.000000
50%,106.000000,3.000000,5.000000,0.370000,45.000000
75%,175.000000,5.000000,24.000000,1.580000,227.000000
max,10000.000000,1250.000000,629.000000,58.500000,365.000000


5. Exploratory Data Analysis

5.1 Average Price by Room Type

In [40]:
df.groupby("room_type")["price"].mean().sort_values(ascending=False)

,price
room_type,
Entire home/apt,211.810918
Private room,89.809131
Shared room,70.248705


In [56]:
df.groupby("neighbourhood_group")["price"].mean().sort_values(ascending=False)

,price
neighbourhood_group,
Manhattan,196.884903
Brooklyn,124.438915
Staten Island,114.812332
Queens,99.517649
Bronx,87.577064


5.2 Average Price by Neighbourhood Group

In [42]:
df.groupby(["neighbourhood_group", "room_type"])["price"].mean().round(2)

neighbourhood_group  room_type      
Bronx                Entire home/apt    127.51
                     Private room        66.89
                     Shared room         59.80
Brooklyn             Entire home/apt    178.35
                     Private room        76.55
                     Shared room         50.77
Manhattan            Entire home/apt    249.26
                     Private room       116.78
                     Shared room         88.98
Queens               Entire home/apt    147.05
                     Private room        71.76
                     Shared room         69.02
Staten Island        Entire home/apt    173.85
                     Private room        62.29
                     Shared room         57.44
Name: price, dtype: float64

5.3 Average Price by Room Type and Neighbourhood Group

In [58]:
df.groupby(["neighbourhood_group", "room_type"])["price"].mean().round(2)

neighbourhood_group  room_type      
Bronx                Entire home/apt    127.51
                     Private room        66.89
                     Shared room         59.80
Brooklyn             Entire home/apt    178.35
                     Private room        76.55
                     Shared room         50.77
Manhattan            Entire home/apt    249.26
                     Private room       116.78
                     Shared room         88.98
Queens               Entire home/apt    147.05
                     Private room        71.76
                     Shared room         69.02
Staten Island        Entire home/apt    173.85
                     Private room        62.29
                     Shared room         57.44
Name: price, dtype: float64

5.4 Top 10 Most Expensive Listings

In [43]:
df.nlargest(10, "price")[[
    "name",
    "neighbourhood_group",
    "neighbourhood",
    "room_type",
    "price"
]]

,name,neighbourhood_group,neighbourhood,room_type,price
9151,Furnished room in Astoria apartment,Queens,Astoria,Private room,10000
17692,Luxury 1 bedroom apt. -stunning Manhattan views,Brooklyn,Greenpoint,Entire home/apt,10000
29238,1-BR Lincoln Center,Manhattan,Upper West Side,Entire home/apt,10000
6530,Spanish Harlem Apt,Manhattan,East Harlem,Entire home/apt,9999
12342,"Quiet, Clean, Lit @ LES & Chinatown",Manhattan,Lower East Side,Private room,9999
40433,2br - The Heart of NYC: Manhattans Lower East ...,Manhattan,Lower East Side,Entire home/apt,9999
30268,Beautiful/Spacious 1 bed luxury flat-TriBeCa/Soho,Manhattan,Tribeca,Entire home/apt,8500
4377,Film Location,Brooklyn,Clinton Hill,Entire home/apt,8000
29662,East 72nd Townhouse by (Hidden by Airbnb),Manhattan,Upper East Side,Entire home/apt,7703
42523,70' Luxury MotorYacht on the Hudson,Manhattan,Battery Park City,Entire home/apt,7500


5.5 Top 10 Neighbourhoods by Number of Listings

In [44]:
df["neighbourhood"].value_counts().head(10)

,count
neighbourhood,
Williamsburg,3919
Bedford-Stuyvesant,3710
Harlem,2658
Bushwick,2462
Upper West Side,1971
Hell's Kitchen,1958
East Village,1853
Upper East Side,1798
Crown Heights,1564


5.6 Listings by Neighbourhood Group

In [45]:
df["neighbourhood_group"].value_counts()

,count
neighbourhood_group,
Manhattan,21660
Brooklyn,20095
Queens,5666
Bronx,1090
Staten Island,373


5.7 Average Minimum Nights by Room Type

In [59]:
df.groupby("room_type")["minimum_nights"].mean().round(2).sort_values(ascending=False)

,minimum_nights
room_type,
Entire home/apt,8.51
Shared room,6.43
Private room,5.38


5.8 Average Availability by Room Type

In [47]:
df.groupby("room_type")["availability_365"].mean().round(2).sort_values(ascending=False)

,availability_365
room_type,
Shared room,161.87
Entire home/apt,111.93
Private room,111.20


5.9 Average Availability by Room Type




In [48]:
df.groupby("room_type")["number_of_reviews"].mean().round(2).sort_values(ascending=False)

,number_of_reviews
room_type,
Private room,24.10
Entire home/apt,22.84
Shared room,16.62


5.10 Top 10 Neighbourhoods by Average Price

In [49]:
avg_price_neighbourhood = (
    df.groupby("neighbourhood")["price"]
      .mean()
      .sort_values(ascending=False)
      .head(10)
      .round(2)
)

avg_price_neighbourhood

,price
neighbourhood,
Fort Wadsworth,800.00
Woodrow,700.00
Tribeca,490.64
Sea Gate,487.86
Riverdale,442.09
Prince's Bay,409.50
Battery Park City,367.56
Flatiron District,341.92
Randall Manor,336.00


5.11 Price and Number of Reviews Correlation

In [50]:
df[["price", "number_of_reviews"]].corr()

,price,number_of_reviews
price,1.000000,-0.047926
number_of_reviews,-0.047926,1.000000


5.12 Top Hosts by Number of Listings

In [51]:
df.groupby(["host_id", "host_name"])["id"].count().sort_values(ascending=False).head(10)

,,id
host_id,host_name,
219517861,Sonder (NYC),327
107434423,Blueground,232
30283594,Kara,121
137358866,Kazuya,103
16098958,Jeremy & Laura,96
12243051,Sonder,96
61391963,Corporate Housing,91
22541573,Ken,87
200380610,Pranjal,65


5.13 Review Status Distribution

In [53]:
df["has_review"].value_counts()

,count
has_review,
Reviewed,38833
No Review,10051


## Key Findings

### Data Cleaning
- The original dataset contained 48,895 Airbnb listings.
- 11 listings with a price of $0 were removed.
- The cleaned dataset contains 48,884 listings.
- Missing values in `name` and `host_name` were replaced with "Unknown".
- Missing values in `reviews_per_month` were replaced with 0.
- `last_review` was converted from object to datetime format.
- No duplicate rows were found.
- Price outliers were identified using the IQR method and flagged rather than removed.

### Exploratory Data Analysis
- Entire home/apt listings have the highest average price at approximately $211.81.
- Manhattan has the largest number of listings with 21,660 listings.
- Williamsburg has the highest number of listings with 3,919 listings.
- Manhattan has the largest number of listings with 21,660 listings.
- Shared rooms have the highest average availability at approximately 161.87 days.
- Private rooms have the highest average number of reviews at approximately 24.10.
- Sonder (NYC) has the highest number of listings among the hosts analyzed, with 327 listings.
- Approximately 79.4% of listings have reviews, while 20.6% have no review history.
- The correlation between price and number of reviews is very weak and negative (-0.0479).

### Conclusion

The analysis shows that Airbnb pricing varies considerably by room type and location.
Entire homes and apartments are generally more expensive than private and shared rooms,
while Manhattan has the highest average listing price. The analysis also shows that
listing price and review count have almost no meaningful linear relationship.

## Conclusion

This project demonstrates practical data cleaning and exploratory data analysis using Python, Pandas, and NumPy.

The dataset was inspected, cleaned, validated, and analyzed to identify patterns in Airbnb listings, pricing, availability, reviews, room types, and locations.

The final cleaned dataset contains 48,884 records and 18 columns.

In [54]:
df.shape

(48884, 18)

CSV Save

In [60]:
df.to_csv("cleaned_airbnb.csv", index=False)